[Orc-support-01] Add dependent classes by Orc from Spark2.3

lee-lei opened the pull request:
## What changes were proposed in this pull request?

There are two kinds of depdendent classes by orc.

The first are used by orc columnar batch reader. Columnar batch
and column vector are changed a lot since Spark2.2. The dependent
classes are directly imported and not changed as follows:

OrcColumnVector.java
ColumnVector.java
ColumnVectorUtils.java
ColumnarArray.java
ColumnarBatch.java
ColumnarMap.java
ColumnarRow.java
Dictionary.java
MutableColumnarRow.java
OffHeapColumnVector.java
OnHeapColumnVector.java
WritableColumnVector.java
UTF8String.java
ArrayData.scala
UnsafeArrayData.java
GenericArrayData.scala

The second are used for deserialization by orc map reduce record reader.
They are also directly imported from Spark2.3 and not changed as follows:

OrcDeserializer.scala
OrcFilters.scala
OrcUtils.scala

Below dependent classes by orc columnar batch reader imported from Spark2.3
are common for Spark2.1 and Spark2.2. Thus add one common directory to put
these dependent classes.

OrcColumnVector.java
ColumnVector.java
ColumnVectorUtils.java
ColumnarArray.java
ColumnarBatch.java
ColumnarMap.java
ColumnarRow.java
Dictionary.java
MutableColumnarRow.java
OffHeapColumnVector.java
OnHeapColumnVector.java
WritableColumnVector.java
UTF8String.java
OrcDeserializer.scala
OrcFilters.scala
OrcUtils.scala

Below three dependent classes are Spark2.1 specific, because Spark2.2 has them.

ArrayData.scala
UnsafeArrayData.java
GenericArrayData.scala

The Spark2.3 has all the above classes, thus doesn't need them.


The code changes are below two categories:

 1.   Add the required dependencies in pom.xml to import orc upstream package.
       And the change is following what Spark did to import orc with one minor change:

         Exclude to import guava package from orc upstream instead of using the
          existing guava package in oap, because oap is using higher version than orc.
   
2.    Add three orc configurations in OapConf.scala used by dependent classess. They are
       also imported from Spark2.3.

       ORC_VECTORIZED_READER_ENABLED
       COLUMN_VECTOR_OFFHEAP_ENABLED
       ORC_COPY_BATCH_TO_SPARK
## How was this patch tested?

The oap unit test passed.

jerrychenhf commented:
Go through the code structure, the following need to be reconsidered:
1. Spark 2.1 has it's own version of ColumnVector at package: java/org/apache/spark/sql/execution/vectorized. If we overwrite the ColumnVector with Spark 2.3 version, the impact to the existing code is unknown. Based on this understanding, we may need to use a different package path for these new classes from Spark 2.3
2.  Based on the fact if we use different package path for these new classes, we don't need to put these class in version specific folders like src/main/spark 2.1/ or src/main/spark 2.2. We can just put these classes in common package in OAP folder and reference them in both 2.1 and 2.2/even 2.3 unchanged.

LuciferYang commented:
Agree with @jerrychenhf，we should impl this features depending on different versions of XXXColumnVector and so on,  copy XXXColumnVector from spark 2.3 and override other versions directly a little violent...

LuciferYang commented on src/main/scala/org/apache/spark/sql/internal/oap/OapConf.scala:
```diff
@@ -245,4 +245,25 @@ object OapConf {
       .stringConf
       .createWithDefault("")
+
+  val ORC_VECTORIZED_READER_ENABLED =
```
Will it conflict with the spark 2.3 sqlconf?

gczsjdy commented on pom.xml:
```diff
@@ -570,7 +624,7 @@
         <scala.version>2.11.8</scala.version>
         <scala.binary.version>2.11</scala.binary.version>
-        <java.version>1.7</java.version>
+        <java.version>1.8</java.version>
```
Should we modify things like `IntStream` to coordinate with java 7?

LuciferYang replied:
`Stream` is just a grammatical sugar. In Spark 2.2 with jdk8+ we can use it, but in Spark 2.1 with jdk7 , I think we shuld rewritte it use old grammatical

gczsjdy replied:
seems applicable @lee-lei

lee-lei replied:
Sure. No problem. Thanks for all of your comments.

jerrychenhf commented:
Just add a new findings after check the code and discuss with the folks.

Spark 2.1 and Spark 2.2 ColumnVector is under a different package path with Spark 2.3 ColumnVector class.
So add the 2.3 ColumnVector class for ORC usage has no problem. Please just make sure these added class will not override any other core classes.

As to whether put these classes in version specific folder or common folder, I just let the engineer folks to decide based on convenience.

lee-lei commented:
By default, the columnar batch and column vector from Spark2.3 have different package path with them from Spark2.1. They will not be overrided. @LuciferYang @jerrychenhf Thanks.

lee-lei commented:
@carsonwang Please help review this PR when you are available. This PR is primarily importing the columnar batch and column vector classes from Spark2.3. The two minor changes are adding the configurations in OapConf.scala and excluding guava from orc upstream package in pom.xml. Thanks.

lee-lei commented:
@JkSelf @gczsjdy @zhixingheyi-tian Please help review this PR when you have free time. Thanks so much!

JkSelf commented:
Thanks @lee-lei  LGTM

zhixingheyi-tian commented:
@lee-lei 
Use the following command

Diff ./OAP/src/main/common-spark2.1-2.2/java/org/apache/spark/sql/vectorized/ColumnarBatch.java ./OAP/src/main/spark2.3/java/org/apache/ Spark/sql/vectorized/ColumnarBatch.java
23c23

The result is:
23c23
< import org.apache.spark.sql.vectorized.MutableColumnarRow;
---
> import org.apache.spark.sql.execution.vectorized.MutableColumnarRow;


The two file codes are almost the same, is there a way to reuse a code file?

gczsjdy commented:
@jerrychenhf @lee-lei For a certain OAP version, it we intend to support k Spark versions and we merge common Spark code changes like this PR did, there will be at most C<sup>1</sup><sub>k</sub>(like spark 2.1, spark2.2, spark 2.3) + C<sup>2</sup><sub>k</sub>(like spark 2.1&2.2, spark2.2&2.3, spark 2.1&2.3) + ...C<sup>k</sup><sub>k</sub>(spark 2.1&2.2&2.3) different repos.
If we take use of this fashion, we should try to make sure the six cases(with k=3) are all merged to one if possible, which is a considerable effort. Also, the directory structure will look bad.

lee-lei commented:
@zhixingheyi-tian The two class are in two different profiles. For OAP case, these two profiles are mutually exclusive. It may not be easy to share one specific class for these two different exclusive profiles. Thanks.

lee-lei commented:
@gczsjdy Thanks for generalization. For ORC support case in OAP for multiple Spark versions, there are C(1, k), C(2, k) two cases. There are not C(3, k), ... C(k, k) cases. The case is for the common back ported classes and utilities from the newer version m. For each back ports from the newer version, there should be at most one common directory. The back ports should be not many. Thanks.

gczsjdy commented:
for k = 3, there will totally be C(1, 3) + C(2, 3) + C(3, 3) = 7 cases.
C(3, 3) means the original src/main
There will be at most 7 folders. We should carefully diff each 2 versions, also diff among 3 version.
And if we don't take use of this fashion, we can reduce to 4 folders(src/main, src/main/spark2.1, src/main/spark2.2, src/main/spark2.3)

lee-lei commented:
@gczsjdy I agree with your conclusion based on your generalization. But I think it needs to be careful and serious to generalize the back port case to your assumption. To be honest, for unifying the common things in the old versions from back ports from the newer version, it's expected to be one common directory. Otherwise, this is not what we want to achieve as you speculated as the above.

zhixingheyi-tian commented:
@lee-lei 
When you specify "Pspark-2.3", will there be two almost identical codes?
If the package is the same as "spark-2.3", can this be solved?

Thanks

lee-lei commented:
@zhixingheyi-tian Sorry. I can't understand your point. Would you please describe a little bit more details? Thanks.

LuciferYang commented:
```
I am afraid introducing a common-2.1and2.2 will introduce much complexity... The directory structure will be unclear and ugly. Using 2 separate folders will actually save efforts.
```
Perhaps we need to reconsider 'how to support multi spark versions' , current scheme a little complicated....

gczsjdy commented:
We came up with a new approach:
- Put version common(can be Spark 2.1&2.2 / Spark 2.2&2.3 / Spark 2.1&2.2&2.3) code in root source directory `/src/main/`
- Put version specific code in seperate source directory(like `/src/main/spark2.1`)

Then no matter when maven tries to build OAP for a specific version or when we want to find a certain Spark code change, the target code should be:

src/main/CERTAIN_VERSION/path/to/code subtract src/main/path/to/code

lee-lei commented:
@gczsjdy I am confused now. The approach you mentioned is what I did in the previous commit. 
@jerrychenhf I assume that the solution you proposed is to put the common-2.1-2.2 to the subdirectory of spark2.3 in order to keep the concise top level directories, since it's directly imported from this version. Would you please help clarify it? Thanks so much!

lee-lei commented:
BTW, the last commit is to move the common dir for 2.1 and 2.2 to the sub dir of spark2.3 as the new solution proposed by Haifeng this morning. Please correct me if anything is wrong. Thanks.

lee-lei commented:
@jerrychenhf @carsonwang @JkSelf @gczsjdy @zhixingheyi-tian If no comments, I will ask Carson to help merge this PR. Thanks so much for your time to reivew!

lee-lei commented on src/main/spark2.3/common-spark2.1-2.2/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java:
```diff
@@ -0,0 +1,193 @@
+/*
```
This orc class is imported from Spark2.3 as well. All the orc classes are also used by both 2.1 and 2.2. So it's expected to put it under the common directory. This class is mentioned in the description of this PR. Thanks.

carsonwang commented:
Thanks all! merging this.

gczsjdy commented on pom.xml:
```diff
@@ -557,8 +611,8 @@
     </plugins>
-    
   </build>
+
```
:nit empty line

lee-lei replied:
Will remove this line in the next PR. Thanks.

LuciferYang commented:
after merge this patch and build with cmd `mvn clean install -Pspark-2.3`, there will compile as follow:
```
[ERROR] /Users/baidu/SourceCode/git/oap-mine/src/main/spark2.3/java/org/apache/spark/sql/vectorized/ColumnarBatch.java:31: ColumnarBatch is already defined as object ColumnarBatch
[ERROR] public final class ColumnarBatch {
[ERROR]                    ^
[ERROR] one error found
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 49.487 s
[INFO] Finished at: 2018-09-07T15:59:24+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.2:compile (scala-compile-first) on project oap: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:3.2.2:compile failed.: CompileFailed -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

lee-lei commented:
@LuciferYang Thanks for the reminder! I will talk with Xiangxiang to resolve this issue. Thanks.

lee-lei commented:
@linhong-intel  We argeed below solution this morning based on the below facts.

These dependent classes are categorized into four types.

1.      Below eight classes are different package path in Spark 2.3 with Spark 2.1 and 2.2. We will keep the same path with 2.3 just for back port.

ColumnVector.java
ColumnVectorUtils.java
ColumnarArray.java
ColumnarBatch.java
ColumnarMap.java
ColumnarRow.java
OffHeapColumnVector.java
OnHeapColumnVector.java
2.      Below seven classes are newly added in Spark 2.3. Spark 2.1 and 2.2 don’t have them.
OrcColumnVector.java
OrcDeserializer.scala
OrcFilters.scala
OrcUtils.scala
MutableColumnarRow.java
Dictionary.java
WritableColumnVector.java

3.   Below one classes are the same package path in Spark 2.3 with 2.1 and 2.2. But Spark 2.3 only adds additional methods in these classes. Except these new added methods, the remaining are the same as Spark 2.1 and 2.2.
UTF8String.java

4.  Below three classes are the same in Spark 2.3 and 2.2. They also just added additional methods based on 2.1. Thus I just replaced these old classes in 2.1 profile with the new classes from 2.2 and 2.3. And they don't need to be put the common directories. The 2.2 and 2.3 will direclty use them from Spark 2.2 and Spark 2.3 package.
ArrayData.scala
GenericArrayData.scala
UnsafeArrayData.java

Based on the above facts, I just added the new commit this noon to distribute all the above common classes in the src/main/java and src/main/scala directories. Thanks.

lee-lei commented:
@linhong-intel  Please check the details in the latest commit of this PR. 
https://github.com/Intel-bigdata/OAP/pull/843/commits/f93f0be18f09832f6421b99bdcc3893ba0fc97b8
Thanks.
